### PR TITLE
Port commitment scheme for HotShot types from Espresso sequencer

### LINKED ISSUE #22: Add commitment functions for HotShot data types (`Header`, etc)
We will need these to implement the batch queue checks on the batch `Justification`

### PR DESCRIPTION
**Pull request by jbearer**
Closes #22

### CODE REVIEW

File: op-service/espresso/commit.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+// WARNING: The string `s` must be a constant. This function does not encode the length of `s` in
+// the hash, which can lead to domain collisions when different strings with different lengths are
+// used depending on the input object.
+func (b *RawCommitmentBuilder) ConstantString(s string) *RawCommitmentBuilder {
+	if _, err := io.WriteString(b.hasher, s); err != nil {
+		panic(fmt.Sprintf("KeccakState Writer is not supposed to fail, but it did: %v", err))
+	}
+
+	// To denote the end of the string and act as a domain separator, include a byte sequence which
+	// can never appear in a valid UTF-8 string.
```
Comment (nomaxg) [inline op-service/espresso/commit.go:41]: Can we add a utf8 check if we are doing this domain separation?

File: op-service/espresso/types_test.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	// Check decoding
+	var decoded Header
+	if err := json.Unmarshal(data, &decoded); err != nil {
+		t.Fatalf("Failed to unmarshal JSON: %v", err)
+	}
+	require.Equal(t, decoded, ReferenceHeader)
+}
+
+// Commitment tests ported from the reference sequencer implementation
+// (https://github.com/EspressoSystems/espresso-sequencer/blob/main/sequencer/src/block.rs)
```
Comment (nomaxg) [inline op-service/espresso/types_test.go:103]: Nice, eventually we can put some KAT vectors in the ```commit``` library for other implementations

**Review by nomaxg — APPROVED** [review]
LGTM, just one small comment

File: op-service/espresso/types_test.go (hunk 2 above)
Comment (jbearer) [reply op-service/espresso/types_test.go:103]: What is a KAT vector?

File: op-service/espresso/commit.go (hunk 1 above)
Comment (jbearer) [reply op-service/espresso/commit.go:41]: You mean check if the string being hashed is valid UTF-8? Sure, I can do that

File: op-service/espresso/types_test.go (hunk 2 above)
Comment (nomaxg) [reply op-service/espresso/types_test.go:103]: Known-answer test vectors for various test cases. In the case, the commit byte sequence.

Comment (jbearer) [reply op-service/espresso/types_test.go:103]: Oh, never heard the term. Definitely seems like a good thing to publish for serialization/commitment stuff

**Pull request merged**